pkg/solver/endpoint_exposer.go

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package solver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/che-incubator/devworkspace-che-operator/apis/che-controller/v1alpha1"
+	"github.com/che-incubator/devworkspace-che-operator/pkg/defaults"
+	dwo "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	routev1 "github.com/openshift/api/route/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type IngressExposer struct {
+	devWorkspaceID     string
+	baseDomain         string
+	ingressAnnotations map[string]string
+	tlsSecretName      string
+}
+
+type RouteExposer struct {
+	devWorkspaceID       string
+	baseDomain           string
+	tlsSecretKey         string
+	tlsSecretCertificate string
+}
+
+type EndpointInfo struct {
+	order        int
+	machineName  string
+	endpointName string
+	port         int32
+	scheme       string
+	service      *corev1.Service
+}
+
+// This method is used compose the object names (both Kubernetes objects and "objects" within Traefik configuration)
+// representing object endpoints.
+func getEndpointExposingObjectName(machineName string, workspaceID string, port int32, endpointName string) string {
+	if endpointName == "" {
+		return fmt.Sprintf("%s-%s-%d", workspaceID, machineName, port)
```

Comment:
Even on Che Server side we tried to move away from machineName. Better not to introduce new usages. I assume it should be replaced with containerName

Reply:
Well, `machineName` really is a `componentName`. I think it is useful to help users with identifying what the generated service/ingress refers to. Would it help if it was called `componentName`?

Btw. the name comes from here :) https://github.com/devfile/devworkspace-operator/blob/main/apis/controller/v1alpha1/devworkspacerouting_types.go#L31

Reply:
actually, the only component that can be used here is container name.
So, containerName or componentName would be better IMHO